utf8.c

```diff
@@ -1026,68 +1026,240 @@ S_unexpected_non_continuation_text(pTHX_ const U8 * const s,
+C<retlen> isn't C<NULL>) to -1.
+
+But note that 0 may also be returned if S<*s> is a legal NUL character.  This
+means that you have to disambiguate a 0 return.  You can do this by checking
+that the first byte of C<s> is indeed a NUL; or by making sure to always pass a
+non-NULL C<retlen> pointer, and by examining it.
+
+Also note that should you wish to proceed with parsing C<s>, you have no easy
+way of knowing where to start looking in it for the next possible character.
+It would be better to have instead called an equivalent function that provides
+this information; any of the C<utf8_to_uv> series, or C<L</utf8n_to_uvchr>>.
+
+=back
+
+Because of these quirks, C<utf8_to_uvchr_buf> is very difficult to use
+correctly and handle all cases.  Generally, you need to bail out at the first
+failure it finds.
+
+The deprecated C<utf8_uvchr> behaves the same way as C<utf8_to_uvchr_buf> for
+well-formed input, and for the malformations it is capable of finding, but
+doesn't find all of them, and it can read beyond the end of the input buffer,
+which is why it is deprecated.
+
+=item Function where the UTF-8 is B<known> to be valid
+
+C<valid_utf8_to_uvchr> is designed
+to be used where you generated the UTF-8 yourself, so you know it is valid.
+It skips any error checking, assuming the sequence of bytes starting at C<s> is
+encoded as Perl extended UTF-8 (or Perl extended UTF-EBCDIC), reading as many
+bytes along C<s> as necessary, and returning that count in C<*retlen> (if
+C<retlen> is not NULL).
+
+=back
+
+=for apidoc      utf8_to_uv_flags
+=for apidoc_item utf8n_to_uvchr
+
+These functions are generalized versions of C<L</utf8_to_uv>>, where you need
+more control over what UTF-8 sequences are acceptable.  These functions are
+unlikely to be needed except for specialized purposes.
+
+C<utf8n_to_uvchr> is more like a generalization of C<utf8_to_uvchr_buf>, with
+fewer quirks, and a different method of specifying the bytes it is allowed to
+examine in C<s>.  It has a C<curlen> parameter instead of a C<e> parameter,
+so the furthest byte in C<s> it can look at is S<C<s + curlen>>.  Its return
+value is, like C<utf8_to_uvchr_buf>, ambiguous with respect to the NUL and
+REPLACEMENT characters, but the value of C<*retlen> can be relied on (except
+with the C<UTF8_CHECK_ONLY> flag described below) to know where the next
+possible character along C<s> starts, removing that quirk.  Hence, you always
+should use C<*retlen> to determine where the next character in C<s> starts.
+
+These functions have an additional parameter, C<flags>, besides the ones in
+C<utf8_to_uv> and C<utf8_to_uvchr_buf>, which can be used to broaden or
+restrict what is acceptable UTF-8.  C<flags> has the same meaning and behavior
+in both functions.  When C<flags> is 0, these functions accept any
+syntactically valid Perl-extended-UTF-8 sequence.
+
+But if C<flags> contains C<UTF8_DISALLOW_ILLEGAL_INTERCHANGE>, all UTF-8
```

Comment:
I think the various flags would ideally be in a table (which we don't have) or a list, which we do have.

I think a list would be easier to use as a reference than the prose version.

This isn't really specific to this PR though.

Reply:
list made

Reply:
Thank you, I found this much more readable.